server/src/main/java/org/elasticsearch/index/mapper/JsonFieldMapper.java

```diff
@@ -415,12 +417,36 @@ protected void parseCreateField(ParseContext context, List<IndexableField> field
             return;
         }
 
-        if (fieldType().indexOptions() != IndexOptions.NONE || fieldType().stored()) {
-            fields.addAll(fieldParser.parse(context.parser()));
-            createFieldNamesField(context, fields);
-        } else {
+        if (fieldType.indexOptions() == IndexOptions.NONE && !fieldType.stored()) {
             context.parser().skipChildren();
+            return;
+        }
+
+        BytesRef storedValue = null;
+        if (fieldType.stored()) {
+            XContentBuilder builder = XContentFactory.jsonBuilder()
+                .prettyPrint()
+                .copyCurrentStructure(context.parser());
+            storedValue = BytesReference.bytes(builder).toBytesRef();
+            fields.add(new StoredField(fieldType.name(), storedValue));
         }
+
+        if (fieldType().indexOptions() != IndexOptions.NONE) {
+            XContentParser indexedFieldsParser = context.parser();
+
+            // If store is enabled, we've already consumed the content to produce the stored field. Here we
```

Comment:
The alternative here would be to add logic to `JsonFieldParser` to generate the stored field as it's parsed. Although it avoids parsing the input twice as we do here, that approach turned turned out fairly messy and didn't separate out concerns as nicely. It also seems uncommon to enable stored fields, so I wasn't as concerned about the fact we will parse the input twice.